pkg/flow/internal/controller/node_declare_component.go

```diff
@@ -0,0 +1,361 @@
+// Exports returns nil if the managed module does not have exports.
+func (cn *DeclareComponentNode) Exports() component.Exports {
+	cn.exportsMut.RLock()
+	defer cn.exportsMut.RUnlock()
+	return cn.exports
+}
+
+func (cn *DeclareComponentNode) LastUpdateTime() time.Time {
+	return cn.lastUpdateTime.Load()
+}
+
+// setExports is called whenever the managed module updates. e must be the
+// same type as the registered exports type of the managed module.
+func (cn *DeclareComponentNode) setExports(e component.Exports) {
+	// Some components may aggressively reexport values even though no exposed
+	// state has changed. This may be done for components which always supply
+	// exports whenever their arguments are evaluated without tracking internal
+	// state to see if anything actually changed.
+	//
+	// To avoid needlessly reevaluating components we'll ignore unchanged
+	// exports.
+	var changed bool
+
+	cn.exportsMut.Lock()
+	if !reflect.DeepEqual(cn.exports, e) {
+		changed = true
+		cn.exports = e
+	}
+	cn.exportsMut.Unlock()
+
+	if changed {
+		// Inform the controller that we have new exports.
+		cn.lastUpdateTime.Store(time.Now())
+		cn.OnComponentUpdate(cn)
+	}
+}
+
+// CurrentHealth returns the current health of the DeclareComponentNode.
+//
+// The health of a DeclareComponentNode is determined by combining:
+//
+//  1. Health from the call to Run().
+//  2. Health from the last call to Evaluate().
+//  3. Health reported from the module.
+func (cn *DeclareComponentNode) CurrentHealth() component.Health {
+	cn.healthMut.RLock()
+	defer cn.healthMut.RUnlock()
+	return component.LeastHealthy(cn.runHealth, cn.evalHealth, cn.managed.CurrentHealth())
+}
+
+// TODO implement debugInfo?
+func (cn *DeclareComponentNode) DebugInfo() interface{} {
+	cn.mut.RLock()
+	defer cn.mut.RUnlock()
+	return nil
+}
+
+// setEvalHealth sets the internal health from a call to Evaluate. See Health
+// for information on how overall health is calculated.
+func (cn *DeclareComponentNode) setEvalHealth(t component.HealthType, msg string) {
+	cn.healthMut.Lock()
+	defer cn.healthMut.Unlock()
+
+	cn.evalHealth = component.Health{
+		Health:     t,
+		Message:    msg,
+		UpdateTime: time.Now(),
+	}
+}
+
+// setRunHealth sets the internal health from a call to Run. See Health for
+// information on how overall health is calculated.
+func (cn *DeclareComponentNode) setRunHealth(t component.HealthType, msg string) {
+	cn.healthMut.Lock()
+	defer cn.healthMut.Unlock()
+
+	cn.runHealth = component.Health{
+		Health:     t,
+		Message:    msg,
+		UpdateTime: time.Now(),
+	}
+}
+
+// ModuleIDs returns the current list of modules that this component is
+// managing.
+func (cn *DeclareComponentNode) ModuleIDs() []string {
+	return cn.moduleController.ModuleIDs()
+}
+
+// BlockName returns the name of the block.
+func (cn *DeclareComponentNode) BlockName() string {
+	return cn.componentName
+}
+
+// This node does not manage any component.
```

Comment:
For consistency, I wish it did; it feels important that anything claiming to be a ComponentNode has a non-nil return value for its managed component instance. 

It's not a difficult interface to implmement (only two methods) so it might be worth doing here. 

Alternatively, we can change the interface to remove this method from component node and make it unique to native components. However, I would still prefer that this return a non-nil value.